### PR TITLE
Improve meta and build infrastructure

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Mostly maintenance update, like getting rid of deprecated dependecies
- set proper conda package dependencies
- clean up CMakeList.txt
- switch to non vendor versioneer
- remove runtime dependency checks
- add toml formatter to pre-commit
- move all available settings from `setup.py` to `pyproject.toml`
- switch from deprecated `python setup.py` to `pypa/build`
- check gdb version only at major

```
conda search local::numba-dpex --info
Loading channels: done
numba-dpex 0.22.1 py310h852cfe9_194
-----------------------------------
file name   : numba-dpex-0.22.1-py310h852cfe9_194.tar.bz2
name        : numba-dpex
version     : 0.22.1
build       : py310h852cfe9_194
build number: 194
size        : 456 KB
license     : Apache-2.0
subdir      : linux-64
url         : file:///home/yevhenii/.miniconda3/conda-bld/linux-64/numba-dpex-0.22.1-py310h852cfe9_194.tar.bz2
md5         : addace56b341db82cd495e0d7f66f87f
timestamp   : 2024-03-28 20:14:34 UTC
dependencies: 
  - __glibc >=2.28,<3.0.a0
  - dpcpp-cpp-rt >=2024.1,<2025.0a0
  - dpcpp-llvm-spirv >=2024.1,<2025.0a0
  - dpctl >=0.17.0dev0,<0.18.0a0
  - dpnp >=0.15.0dev1,<0.16.0a0
  - intel-cmplr-lib-rt >=2024.1,<2025.0a0
  - libgcc-ng >=13
  - libstdcxx-ng >=13
  - llvmlite >=0.42.0,<0.43.0a0
  - numba >=0.59.1,<0.60.0a0
  - numpy >=1.26.4,<2.0a0
  - python >=3.10,<3.11.0a0
  - python_abi 3.10.* *_cp310
```

Checklist
- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: .github/actions/setup-miniconda/action.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    - name: Setup conda
+      if: steps.cache.outputs.cache-hit == 'true'
+      uses: conda-incubator/setup-miniconda@v3
+      with:
+        activate-environment: ${{inputs.activate-environment}}
+        channels: ${{ inputs.channels }}
+        environment-file: ${{inputs.activate-environment}}.lock
+        run-post: ${{ inputs.run-post }}
+
+    - name: Disable defautls
```
Comment (oleksandr-pavlyk) [inline .github/actions/setup-miniconda/action.yml:94]:
```suggestion
    - name: Disable defaults
```

File: .github/workflows/conda-package.yml [hunk 2]
```diff
@@ -176,14 +169,8 @@ jobs:
 
       # We want to make sure that all dependecies install automatically.
       # intel::intel-opencl-rt is needed for set-intel-ocl-icd-registry.ps1
-      # TODO: remove pytest dependency directly, once resolved:
-      # https://github.com/pytest-dev/pytest/issues/12036
       - name: Install builded package
```
Comment (oleksandr-pavlyk) [inline .github/workflows/conda-package.yml:172]:
```suggestion
      - name: Install built package
```

File: conda-recipe/meta.yaml [hunk 3]
```diff
@@ -27,29 +28,41 @@ requirements:
...
-        - dpnp >=0.14*
-        - dpcpp-llvm-spirv
-        - wheel
+        - setuptools >=69
+        - scikit-build >=0.17
+        - ninja >=1.11.1 # [not win]
+        - cmake >=3.29
+        - numba >=0.59
+        - llvmlite >=0.42.0
+        - dpctl >=0.16
```
Comment (diptorupd) [inline conda-recipe/meta.yaml:37]:
```suggestion
        - dpctl >0.16
```
If the code path to support version 0.16 is removed from `dpctl_iface._herlper.py`, then latest numba-dpex will no longer be compatible with 0.16 and lower versions of dpctl.

File: conda-recipe/meta.yaml [hunk 4]
```diff
@@ -27,29 +28,41 @@ requirements:
...
+        - wheel >=0.43
+        - pip >=24.0
+        - python-build >=1.1
+        - versioneer==0.29
+        # versioneer dependency
+        - tomli # [py<311]
     run:
         - {{ pin_compatible('dpcpp-cpp-rt', min_pin='x.x', max_pin='x') }}
+        - {{ pin_compatible('intel-cmplr-lib-rt', min_pin='x.x', max_pin='x') }}
+        - {{ pin_compatible('dpcpp-llvm-spirv', min_pin='x.x', max_pin='x') }}
```
Comment (diptorupd) [inline conda-recipe/meta.yaml:52]: what does `x.x` imply?

File: numba_dpex/__init__.py
Comment (diptorupd) [inline numba_dpex/__init__.py:1]: What is the motivation behind the change to remove semantic version check of dpctl and numba? If we go these way, then documentation needs to be updated that the minimum supported dpctl is 0.16.1 and numba is 0.59.

Comment (ZzEeKkAa) [reply numba_dpex/__init__.py:1]: Motivation is to reduce codebase and make maintenance easier. Project is not suppose to cover all the available combination of dependent libraries. It is not the goal and even if it would be, there is no benefit from it I can imagine.

File: conda-recipe/meta.yaml (hunk 4 above)
Comment (ZzEeKkAa) [reply conda-recipe/meta.yaml:52]: This particular line means to support dpcpp from 2024.1 to 2025. You can read more about it here: https://docs.conda.io/projects/conda-build/en/stable/resources/variants.html#pinning-at-the-recipe-level
This is the same approach dpctl and dpnp uses

File: conda-recipe/meta.yaml (hunk 3 above)
Comment (ZzEeKkAa) [reply conda-recipe/meta.yaml:37]: Yes, that's right. I've thought those changes will be included in  0.16.1. If that is true then `dpctl>0.16.1` must take place. Please confirm what is the desired version.

File: conda-recipe/meta.yaml (hunk 4 above)
Comment (ZzEeKkAa) [reply conda-recipe/meta.yaml:52]: This is how dependencies look like after those resolutions:
```
dependencies: 
  - __glibc >=2.28,<3.0.a0
  - dpcpp-cpp-rt >=2024.1,<2025.0a0
  - dpcpp-llvm-spirv >=2024.1,<2025.0a0
  - dpctl >=0.17.0dev0,<0.18.0a0
  - dpnp >=0.15.0dev1,<0.16.0a0
  - intel-cmplr-lib-rt >=2024.1,<2025.0a0
  - libgcc-ng >=13
  - libstdcxx-ng >=13
  - llvmlite >=0.42.0,<0.43.0a0
  - numba >=0.59.1,<0.60.0a0
  - numpy >=1.26.4,<2.0a0
  - python >=3.10,<3.11.0a0
  - python_abi 3.10.* *_cp310
```

**Review by diptorupd — DISMISSED** [review]
Nice changes!

**Review by diptorupd — APPROVED** [review]
